File: Week 7 - 24.03.2021 DAA Lab/0-1 Knapsack using Dynamic Programming_DAY_7.cpp
// Author : Abhishek Sharma
// Knapsack Problem using C++ language
// difficulty : medium
// score : 10



#include <bits/stdc++.h>
using namespace std;
int max(int a, int b) 
{ 
  return (a > b) ? a : b; 
}
// pythonpapa
int knapSack(int W, int wt[], int val[], int n)
{

    if (n == 0 || W == 0)
        return 0;

    if (wt[n - 1] > W)
        return knapSack(W, wt, val, n - 1);

    else
        return max(
            val[n - 1]
                + knapSack(W - wt[n - 1], 
                        wt, val, n - 1),
            knapSack(W, wt, val, n - 1));
}

// Author : Abhishek Sharma
int main()
{
    int n, W, wt[10], val[10];
    cin>>n;
    cin>>W;
    for (int i = 0; i<n; i++){
        cin>>wt[i];
    }
    for (int i = 0; i<n; i++){
        cin>>val[i];
    }
    // pythonpapa
    cout << knapSack(W, wt, val, n);
    return 0;
}
// Author : Abhishek Sharma

File: Week 7 - 24.03.2021 DAA Lab/0-1 Knapsack using Dynamic Programming_DAY_7.java
// Author : Abhishek Sharma
// Knapsack problem using the JAVA
// difficulty : medium
// score : 10


import java.util.*;
// Author : Abhishe SHarma
public class Solution {
    static int max(int a, int b) 
    { 
    return (a > b) ? a : b; 
    }
    // pythonpapa
    static int knapSack(int W, int wt[], int val[], int n)
    {
        // Base Case
        if (n == 0 || W == 0)
            return 0;
        if (wt[n - 1] > W)
            return knapSack(W, wt, val, n - 1);
            // pythonpapa
        else
            return max(val[n - 1]
                    + knapSack(W - wt[n - 1], wt,
                                val, n - 1),
                    knapSack(W, wt, val, n - 1));
    }
  // Pythonpapa
    public static void main(String args[])
    {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        int W = sc.nextInt();
        int[] wt = new int[10];
        int[] val = new int[10];
        for (int i=0; i<n;i++){
            wt[i] = sc.nextInt();
        }
        for (int i=0; i<n;i++){
            val[i] = sc.nextInt();
        }
        System.out.println(knapSack(W, wt, val, n));
    }
}
// Code by Abhishek Sharma, 2021
